xdsl/interactive/app.py

```diff
@@ -0,0 +1,140 @@
+"""
+This app allows you to paste xDSL IR into the Input TextArea, select a pass (or multiple passes) to be applied on the input IR
+and subsequently the IR generated by the application of the selected pass(es) is displayed in the Output TextArea. The selected
+passes are displayed in the top right "Selecred passes/query" box. The "Condense" button filters the pass selection list to
+display passes that change the IR, or require arguments to be executed (and thus "may" change the IR). The "Uncondense" button
+returns the selection list to contain all the passes. The "Clear Passes" Button removes the application of the selected passes.
+The "Copy Query" Button allows the user to copy the selected passes/query that they have so far selected (i.e. copy the top right
+box).
+
+This app is still under construction.
+"""
+
+from io import StringIO
+
+from rich.style import Style
+from textual import events, on
+from textual.app import App, ComposeResult
+from textual.containers import Horizontal, Vertical
+from textual.reactive import reactive
+from textual.widgets import Footer, TextArea
+from textual.widgets.text_area import TextAreaTheme
+
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import MLContext
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.tools.command_line_tool import get_all_dialects
+
+
+class OutputTextArea(TextArea):
+    """Used to prevent users from being able to change/alter the Output TextArea"""
+
+    async def _on_key(self, event: events.Key) -> None:
+        event.prevent_default()
+
+
+class InputApp(App[None]):
+    """
+    Class buildling the Interactive Compilation App. Uses Textual Python to construct reactive variables on the App, Event,
+    Widget classes etc. The UI is derived from those values as they change.
+    """
+
+    CSS_PATH = "app.tcss"
+
+    BINDINGS = [
+        ("d", "toggle_dark", "Toggle dark mode"),
+        ("q", "quit_app", "Quit"),
+    ]
+
+    current_module: reactive[ModuleOp | Exception | None] = reactive(None)
+    """
+    Reactive variable used to save the current state of the modified Input TextArea (i.e. is the Output TextArea)
+    """
+
+    input_text_area = TextArea(id="input")
+    output_text_area = OutputTextArea(id="output")
+
+    def compose(self) -> ComposeResult:
+        """
+        Creates the required widgets, events, etc.
+        Get the list of xDSL passes, add them to an array in "Selection" format (so it can be added to a Selection List)
+        and sort the list in alphabetical order.
+        """
+
+        # defines a theme for the Input/Output TextArea's
+        my_theme = TextAreaTheme(
```

Comment:
This constant makes sense at the top level. In python, the convension is `UPPER_SNAKE_CASE` for constants. I would recommend calling it `_DEFAULT_TEXT_AREA_THEME` at the top level.